src/mpl/include/mpl_thread_argobots.h

```diff
@@ -52,9 +52,29 @@ typedef void (*MPL_thread_func_t) (void *data);
 void MPL_thread_create(MPL_thread_func_t func, void *data, MPL_thread_id_t * idp, int *errp);
 
 #define MPL_thread_exit()
-#define MPL_thread_self(idp_) ABT_thread_self(idp_)
-#define MPL_thread_join(id_) ABT_thread_free(&id_)
-#define MPL_thread_same(idp1_, idp2_, same_)  ABT_thread_equal(*idp1_, *idp2_, same_)
+#define MPL_thread_self(idp_)                                                 \
+    do {                                                                      \
+        ABT_thread self_thread_tmp_ = ABT_THREAD_NULL;                        \
+        ABT_thread_self(&self_thread_tmp_);                                   \
+        if (self_thread_tmp_ == ABT_THREAD_NULL) {                            \
+            /* It seems that an external thread calls this function. */       \
+            /* Use Pthreads ID instead. */                                    \
+            *(idp_) = (uintptr_t)pthread_self();                              \
+        } else {                                                              \
+            *(idp_) = (uintptr_t)self_thread_tmp_;                            \
+        }                                                                     \
```

Comment:
Can we tell from the id that whether an id is an external thread id or an argobot thread id? For example, can we apply a mask bit to the id? That may mean that id for pthread returned here is not the same as the original pthread_id, but I think as long as it is consistent, it should be fine.

Reply:
Thanks. Yes, what this PR actually does is extending the Argobots version `MPL_thread_id_t` so that it can accommodate both Argobots thread ID and Pthreads ID. The purpose of this PR is to support external threads when MPICH is configured with Argobots. Let me update the title of this PR once we figure out how this PR will be.

> Can we tell from the id that whether an id is an external thread id or an argobot thread id?

Currently no, but the current MPICH does not require it. If `MPL_thread_id_t` should have this feature, there are a few options, but I feel none of them are nice.

1. Using `struct`:
  I originally implemented the Argobots-version `MPL_thread_id_t` as a struct, but I found that code assumes that `MPL_thread_id_t` is an integral type (like `int` or `void *`) and therefore assigns `0` to `MPL_thread_id_t`. Without changing the existing code, it is hard. I am not sure if we should change the existing code base just for this PR.

2. Using `bit-mask`:
  Argobots thread ID (in this case `ABT_thread`) does not use the rightmost 2 or 3 bits since I know Argobots internally uses a `malloc()`'ed pointer address for the ID which should be properly aligned with 4 or 8 bytes. I am not sure if we can do the same thing for Pthreads. For example, Solaris' `pthread_t` seems 4 bytes even on 64-bit machine when I checked -- obviously this `pthread_t` is not an address. Finding unused bits in `pthread_t` is not easy.

3. Using a hash table
  I can internally create a hash table that maps `MPL_thread_id_t` to either `pthread_t` or `ABT_thread`, but it is quite complicated.

4. Using a 16-byte integer for `MPL_thread_id_t`.
  I am not sure if we should rely on a non-portable 16-byte integer, but if we use this, we can have enough unused bits. `pthread_t` is either 4 or 8 bytes on virtually all the platforms (as far as I checked Mac/Linux/FreeBSD/Solaris/Cygwin)

Reply:
>     2\. Using `bit-mask`:
>         Argobots thread ID (in this case `ABT_thread`) does not use the rightmost 2 or 3 bits since I know Argobots internally uses a `malloc()`'ed pointer address for the ID which should be properly aligned with 4 or 8 bytes. I am not sure if we can do the same thing for Pthreads. For example, Solaris' `pthread_t` seems 4 bytes even on 64-bit machine when I checked -- obviously this `pthread_t` is not an address. Finding unused bits in `pthread_t` is not easy.

Since the purpose of this id is for MPICH to internally tell the threads apart and the id is never fed back to `pthread` or `argobot`, there isn't need to preserve the original id. For example, let's say we "blindly" always set the highest bit to 1 for argobot thread id and always set it to 0 for posix thread id, will that work?

Reply:
If there's no conflict, yes this works. I will implement it.

Reply:
I added a special bit to check the thread type.
I also changed the title of this PR and the commit headline.